pkg/controller/jobs/sparkapplication/sparkapplication_controller.go

```diff
@@ -0,0 +1,320 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sparkapplication
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"slices"
+
+	sparkv1beta2 "github.com/kubeflow/spark-operator/v2/api/v1beta2"
+	sparkcommon "github.com/kubeflow/spark-operator/v2/pkg/common"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta2"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/podset"
+)
+
+var (
+	gvk = sparkv1beta2.GroupVersion.WithKind("SparkApplication")
+)
+
+const (
+	FrameworkName      = "sparkoperator.k8s.io/sparkapplication"
+	driverPodSetName   = "driver"
+	executorPodSetName = "executor"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewJob:        NewJob,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupWebhook,
+		JobType:       &sparkv1beta2.SparkApplication{},
+		AddToScheme:   sparkv1beta2.AddToScheme,
+	}))
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update
+// +kubebuilder:rbac:groups=sparkoperator.k8s.io,resources=sparkapplications,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloadpriorityclasses,verbs=get;list;watch
+
+func NewJob() jobframework.GenericJob {
+	return &SparkApplication{}
+}
+
+var NewReconciler = jobframework.NewGenericReconcilerFactory(NewJob)
+
+type SparkApplication sparkv1beta2.SparkApplication
+
+var _ jobframework.GenericJob = (*SparkApplication)(nil)
+
+func (j *SparkApplication) Object() client.Object {
+	return (*sparkv1beta2.SparkApplication)(j)
+}
+
+func (j *SparkApplication) IsSuspended() bool {
+	return j.Spec.Suspend != nil && *j.Spec.Suspend
+}
+
+func (j *SparkApplication) IsActive() bool {
+	return j.Status.AppState.State == sparkv1beta2.ApplicationStateRunning
+}
+
+func (j *SparkApplication) Suspend() {
+	j.Spec.Suspend = ptr.To(true)
+}
+
+func (j *SparkApplication) GVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *SparkApplication) PodLabelSelector() string {
+	return fmt.Sprintf("%s=%s", sparkcommon.LabelSparkAppName, j.Name)
+}
+
+func (j *SparkApplication) PodSets(ctx context.Context) ([]kueue.PodSet, error) {
+	// driver and executor
```

Comment:
I'd like us to check this assumption. 
From what I read and understand, there should only be 1 driver but my AI search says that there can be many executors.

SparkApplications say there can be multiple instances of executors. Would there ever be more than 2 pods for SparkApplications?

cc @andreyvelich

Reply:
Yes, SparkApplication has 1 Driver and multiple Executors. The number of initial executors is set in `.instances` API: https://github.com/kubeflow/spark-operator/blob/master/examples/spark-pi-kube-scheduler.yaml#L44.

Since we don't support `dynamicAllocation` yet, the number of executors is static.

Reply:
Sorry the count means that the executors can be replicated but I guess they will always have the same template. So count in PodSet represent the executor instances.

Reply:
> I guess they will always have the same template
Yes, as far as I know, spark does not support heterogeneous-sized executor pods.

> So count in PodSet represent the executor instances.

I understand the same!